File: prism-sdk/crypto/jvm/src/main/java/io/iohk/atala/prism/crypto/japi/EC.java
package io.iohk.atala.prism.crypto.japi;

import io.iohk.atala.prism.crypto.AndroidEC$;
import io.iohk.atala.prism.crypto.EC$;

import java.math.BigInteger;
import java.util.Arrays;

public interface EC {
    public static EC getInstance(CryptoProvider provider) {
        switch (provider) {
            case JVM:
                return new ECFacade(EC$.MODULE$);
            case Android:
                return new ECFacade(AndroidEC$.MODULE$);
            default:
                throw new IllegalArgumentException(
                        String.format("Unexpected provider %s, available types are %s",
                                provider, Arrays.toString(CryptoProvider.values())));
        }
    }

    /**
     * Generates a P-256k/secp256k1/prime256v1 key-pair.
     */
    ECKeyPair generateKeyPair();

    /**
     * Returns the key-pair represented by the given private key's `D` as byte array.
     */
    ECKeyPair toKeyPairFromPrivateKey(byte[] d);

    /**
     * Returns the key-pair represented by the given private key's `D` as number.
     */
    ECKeyPair toKeyPairFromPrivateKey(BigInteger d);

    /**
     * Returns the private key represented by the given byte array.
     */
    ECPrivateKey toPrivateKey(byte[] d);

    /**
     * Returns the private key represented by the given number.
     */
    ECPrivateKey toPrivateKey(BigInteger d);

    /**
     * Returns the public key represented by the given encoded byte array.
     */
    ECPublicKey toPublicKey(byte[] encoded);

    /**
     * Returns the public key represented by the given coordinates as byte arrays.
     */
    ECPublicKey toPublicKey(byte[] x, byte[] y);

    /**
     * Returns the public key represented by the given coordinates.
     */
    ECPublicKey toPublicKey(BigInteger x, BigInteger y);

    /**
     * Returns the public key represented by the given private key's `D` as byte array.
     */
    ECPublicKey toPublicKeyFromPrivateKey(byte[] d);

    /**
     * Returns the public key represented by the given private key's `D` as number.
     */
    ECPublicKey toPublicKeyFromPrivateKey(BigInteger d);

    /**
     * Signs the given text with the given private key.
     */
    ECSignature sign(String text, ECPrivateKey privateKey);

    /**
     * Signs the given data with the given private key.
     */
    ECSignature sign(byte[] data, ECPrivateKey privateKey);

    /**
     * Verifies whether the given text matches the given signature with the given public key.
     */
    boolean verify(String text, ECPublicKey publicKey, ECSignature signature);

    /**
     * Verifies whether the given data matches the given signature with the given public key.
     */
    boolean verify(byte[] data, ECPublicKey publicKey, ECSignature signature);
}

File: prism-kotlin-sdk/crypto/src/commonMain/kotlin/io/iohk/atala/prism/kotlin/crypto/EC.kt
package io.iohk.atala.prism.kotlin.crypto

import com.ionspin.kotlin.bignum.integer.BigInteger
import io.iohk.atala.prism.kotlin.crypto.keys.ECKeyPair
import io.iohk.atala.prism.kotlin.crypto.keys.ECPoint
import io.iohk.atala.prism.kotlin.crypto.keys.ECPrivateKey
import io.iohk.atala.prism.kotlin.crypto.keys.ECPublicKey
import io.iohk.atala.prism.kotlin.crypto.signature.ECSignature

expect object EC {
    /**
     * Generates a secp256k1 key-pair.
     */
    fun generateKeyPair(): ECKeyPair

    /**
     * Returns the private key represented by the given byte array.
     */
    fun toPrivateKey(encoded: ByteArray): ECPrivateKey

    /**
     * Returns the private key represented by the given number.
     */
    fun toPrivateKey(d: BigInteger): ECPrivateKey

    /**
     * Returns the public key represented by the given encoded byte array.
     */
    fun toPublicKey(encoded: ByteArray): ECPublicKey

    /**
     * Returns the public key represented by the given coordinates as byte arrays.
     */
    fun toPublicKey(x: ByteArray, y: ByteArray): ECPublicKey

    /**
     * Returns the public key represented by the given coordinates.
     */
    fun toPublicKey(x: BigInteger, y: BigInteger): ECPublicKey

    /**
     * Returns the public key represented by the given private key.
     */
    fun toPublicKeyFromPrivateKey(privateKey: ECPrivateKey): ECPublicKey

    /**
     * Returns the signature represented by the given encoded byte array.
     */
    fun toSignature(encoded: ByteArray): ECSignature

    /**
     * Signs the given text with the given private key.
     */
    fun sign(text: String, privateKey: ECPrivateKey): ECSignature

    /**
     * Signs the given data with the given private key.
     */
    fun sign(data: ByteArray, privateKey: ECPrivateKey): ECSignature

    /**
     * Verifies whether the given text matches the given signature with the given public key.
     */
    fun verify(text: String, publicKey: ECPublicKey, signature: ECSignature): Boolean

    /**
     * Verifies whether the given data matches the given signature with the given public key.
     */
    fun verify(data: ByteArray, publicKey: ECPublicKey, signature: ECSignature): Boolean

    /**
     * Verifies whether the given point belongs to the secp256k1 curve.
     */
    fun isSecp256k1(point: ECPoint): Boolean
}
